libbpf-tools/trace_helpers.c

```diff
@@ -1043,7 +1023,38 @@ bool fentry_exists(const char *name, const char *mod)
 		}
 	}
 
-err_out:
+	return id;
+}
+
+bool fentry_exists(const char *name, const char *mod)
+{
+	const char sysfs_vmlinux[] = "/sys/kernel/btf/vmlinux";
+	struct btf *base, *btf = NULL;
+	char sysfs_mod[80];
+	int id = -1;
+
+	base = btf__parse(sysfs_vmlinux, NULL);
+	if (libbpf_get_error(base)) {
+		fprintf(stderr, "failed to parse vmlinux BTF at '%s': %s\n",
+			sysfs_vmlinux, strerror(-libbpf_get_error(base)));
+		goto out;
+	}
+	if (mod) {
+		snprintf(sysfs_mod, sizeof(sysfs_mod), "/sys/kernel/btf/%s", mod);
+		btf = btf__parse_split(sysfs_mod, base);
+		if (libbpf_get_error(btf)) {
+			fprintf(stderr, "failed to load BTF from %s: %s\n",
+				sysfs_mod, strerror(-libbpf_get_error(btf)));
+			goto out;
+		}
+		id = btf_id_by_name(btf, name);
+		if (id > 0)
+			goto out;
```

Comment:
it's unnecessary to do btf_id_by_name() twice. If module BTF was successfully loaded, then iterating it from type #1 to the last one will also iterate vmlinux BTFs (module BTF is "built on top" of vmlinux BTF and includes all the vmlinux BTF types as well). So you can simplify this and you won't need btf_id_by_name() function, you can leave the logic as is. Basically, the idea can be experessed in pseudo code as:

```
base = btf__parse();
if (error) /* exit, no BTF at all */
btf = btf__parse_split(sysfs_mod, base);
if (error)
    btf = base; /* no module, but fallback to plain vmlinux BTF */
/* search for type in btf just once */
```

Reply:
Done.